akka-http-core/src/main/scala/akka/http/Http.scala

```diff
@@ -173,6 +174,217 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
+   * It cannot fully guarantee that the application will always run at most one pool for a given host and
+   * configuration.
+   * The cache removes entries for pools that are shut down (e.g. because of an idle timeout) and due to the
+   * inherent raciness of this operation one client might end up re-starting an old pool while another
+   * client starts a new pool if both access the cache around the time of the shutdown.
+   *
+   * Since the underlying transport usually comprises more than a single connection the produced flow might generate
+   * responses in an order that doesn't directly match the consumed requests.
+   * For example, if two requests A and B enter the flow in that order the response for B might be produced before the
+   * response for A.
+   * In order to allow for easy response-to-request association the flow takes in a custom, opaque context
+   * object of type ``T`` from the application which is emitted together with the corresponding response.
+   */
+  def cachedHostConnectionPool[T](host: String, port: Int = 80,
+                                  options: immutable.Traversable[Inet.SocketOption] = Nil,
+                                  settings: HostConnectionPoolSettings = HostConnectionPoolSettings(system),
+                                  defaultHeaders: immutable.Seq[HttpHeader] = Nil,
+                                  log: LoggingAdapter = system.log)(implicit fm: FlowMaterializer): Flow[(HttpRequest, T), (Try[HttpResponse], T), HostConnectionPool] = {
+    val cps = ConnectionPoolSetup(encrypted = false, options, settings, defaultHeaders, log)
+    val setup = HostConnectionPoolSetup(host, port, cps)
+    cachedHostConnectionPool(setup)
+  }
+
+  // every ActorSystem maintains its own connection pools
+  private[this] val hostPoolCache = new ConcurrentHashMap[HostConnectionPoolSetup, Future[ConnectionPool.Gateway]]
+
+  /**
+   * Returns a [[Flow]] which dispatches incoming HTTP requests to the per-ActorSystem pool of outgoing
+   * HTTP connections to the given target host endpoint. For every ActorSystem, target host and pool
+   * configuration a separate connection pool is maintained.
+   * The HTTP layer transparently manages idle shutdown and restarting of connections pools as configured.
+   * The returned [[Flow]] instances therefore remain valid throughout the lifetime of the application.
+   *
+   * Note that the provided caching logic follows a "best-effort" design with regard to pool uniqueness.
+   * It cannot fully guarantee that the application will always run at most one pool for a given host and
+   * configuration.
+   * The cache removes entries for pools that are shut down (e.g. because of an idle timeout) and due to the
+   * inherent raciness of this operation one client might end up re-starting an old pool while another
+   * client starts a new pool if both access the cache around the time of the shutdown.
```

Comment:
@sirthias what happens to gateways that have been delivered previously (pruned from the cache but still referenced)? They will still need to be able to forward requests to any later created actual pool instance. But that would mean that you need to decouple the things you hand out from the things you put into the cache.